flashlight/fl/tensor/backend/af/CMakeLists.txt

```diff
@@ -58,9 +58,15 @@ if (NOT (${_AF_CUDA} OR ${_AF_CPU} OR ${_AF_OPENCL}))
 
   # Exactly one ArrayFire target is found, but no options are specified.
   # Set the option based on the found target.
-  set(${FL_ARRAYFIRE_USE_CUDA}   TARGET ArrayFire::afcuda)
-  set(${FL_ARRAYFIRE_USE_CPU}    TARGET ArrayFire::afcpu)
-  set(${FL_ARRAYFIRE_USE_OPENCL} TARGET ArrayFire::afopencl)
+  if (TARGET ArrayFire::afcuda)
```

Comment:
Any chance we can use the [`TARGET_EXISTS`](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#target-dependent-expressions) CMake generator expression to make this work? It seems the `TARGET [target]` behavior was UB this whole time and removed by new versions of CMake.

This would look like
```cmake
set(${FL_ARRAYFIRE_USE_OPENCL} $<TARGET_EXISTS:ArrayFire::afcuda>)
```

since I can't repro, say so here if this doesn't work and we can merge anyways.

Reply:
I actually initially tried the same thing, but discovered that cmake generator expressions are only evaluated at the build stage, not the configure stage. So using generator expressions to switch functionality at the configure stage wont work. If you try printing the variable, ie.
```cmake
set(FL_ARRAYFIRE_USE_CUDA $<TARGET_EXISTS:ArrayFire::afcuda>)
message(STATUS "FL_ARRAYFIRE_USE_CUDA: ${FL_ARRAYFIRE_USE_CUDA}")
```
the output is:
`FL_ARRAYFIRE_USE_CUDA: $<TARGET_EXISTS:ArrayFire::afcuda>`

Also I believe 
```cmake
set(${FL_ARRAYFIRE_USE_CUDA} $<TARGET_EXISTS:ArrayFire::afcuda>)
```
is incorrect usage as cmake would evaluate the variable before evaluating the set command. So if the value of `FL_ARRAYFIRE_USE_CUDA` was `OFF`, it would be
```cmake
set(OFF $<TARGET_EXISTS:ArrayFire::afcuda>)
```
Which obviously wont work, but silently fails in cmake.

Reply:
@benborder makes sense! Merging.